proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,314 @@
+======================================
+
+.. author:: Brandon Chinn
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/570>`_.
+.. sectnum::
+.. contents::
+
+Most languages have support for interpolating variables and (usually) arbitrary expressions:
+
+* Python
+
+  .. code-block:: python
+
+    f"Expected: {x + y}, got: {result}"
+
+* Rust
+
+  .. code-block:: rust
+
+    format!("Expected: {sum}, got: {result}")
+
+* Scala
+
+  .. code-block:: scala
+
+    s"Expected: ${x + y}, got: ${result}"
+
+* Javascript/Typescript
+
+  .. code-block:: javascript
+
+    `Expected: ${x + y}, got: ${result}`
+
+This proposal proposes adding S-strings (like Scala's syntax) to Haskell.
+
+Motivation
+----------
+
+Most non-trivial projects build strings at some point: printing out logs, rendering exceptions, generating code, pretty-printing. There are currently multiple ways to do this:
+::
+
+  -- concatenation + show
+  error $ "Expected: " <> show (x + y) <> ", got: " <> show result
+
+  -- printf
+  error $ printf "Expected: %d, got: %d" (x + y) result
+
+  -- safer printf, e.g. the `formatting` package
+  error $ format ("Expected: " % int % ", got: " % int) (x + y) result
+
+  -- quasiquoters, e.g. `string-interpolate` using `haskell-src-exts`
+  error [i|Expected: #{x + y}, got: #{result}|]
+
+But each of these leave things to be desired:
+
+* Manual interpolation (e.g. ``<>``, ``show``, ``unwords``, etc.) is annoying, especially for strings with a lot of interpolation. It's hard to see the overall structure of the string, especially when building up a ``Text``:
+  ::
+
+    let
+      name1 = _ :: Text
+      age1 = _ :: Int
+      name2 = _ :: Text
+      age2 = _ :: Int
+
+      textExample1 = name1 <> " (age: " <> T.pack (show age1) <> ") encountered " <> name2 <> " (age: " <> T.pack (show age2) <> ")"
+
+      textExample2 = T.pack $ T.unpack name1 <> " (age: " <> show age1 <> ") encountered " <> T.unpack name2 <> " (age: " <> show age2 <> ")"
+
+      textExample3 = T.unwords
+        [ name1
+        , "(age: " <> T.pack (show age1) <> ")"
+        , "encountered"
+        , name2
+        , "(age: " <> T.pack (show age2) <> ")"
+        ]
+
+* ``printf`` is partial and unsafe, which especially safety-conscious people might always stay away from anyway. Using a safer ``printf`` like ``formatting`` induces a third-party dependency, which is admittedly lightweight, but isn't as seamless of an integration as native string interpolation would be
```

Comment:
This argument generalises to building everything into GHC and moving every useful library into `base`. People liked Python's included batteries until they proved hard to changes, and now say things like "the standard library is where packages go to die". I'd like to know what the limiting principle is, because we could bundle everything with GHC in the name of convenience.

Perversely, I actually think that the fact that GHC Haskell is weird and limited has paid off with libraries that do so much better than the obvious fix: lenses are better than a naive record dot proposal, `formatting` is a lot better than interpolation, and we should not copy things from other languages just because they are there.

Reply:
> I'd like to know what the limiting principle is, because we could bundle everything with GHC in the name of convenience.

Well in this specific instance, we wouldn't "just" be bundling `formatting` into GHC. The benefit of adding this over just recommending `formatting` is because `formatting` wouldn't be able to take advantage of multiline strings (if that proposal is accepted). Compare:

```hs
let s =
      format
      ("""
      This is an example.
      Here is a list of people and their ages:
      Person named """ % text % " has age " % int % """.
      Another person named """ % text % " has age " % int % """.
      The end.
      """)
      name1 age1 name2 age2
```
vs
```hs
let s =
      s"""
      This is an example.
      Here is a list of people and their ages:
      Person named ${name1} has age ${age1}.
      Another person named ${name2} has age ${age2}.
      The end.
      """
```
Excusing the contrived example for a second, the second has much less glue getting in the way. If this isn't compelling to you, maybe we can just agree to disagree 🤷 

> `formatting` is a lot better than interpolation

I guess I just personally disagree

Reply:
The thing I find compelling about `formatting` is that it looks more scalable to me than interpolation - it is easy to specify additional tweaks to each placeholder, use different functions to interpret the interpolation in different ways, etc. (Similarly, I use lenses over record sugar because it is more scalable).

I am lukewarm on multiline strings, mostly because I don't mind TH as much as some, but I think a straightfoward take on interpolation has the potential to be comfortably limiting.

Have you considered a `formatting`-style interpolation syntax, where `format f"Splice #{text} and #{int}` someText aNumber` -> `format ("Splice " % text % " and " % int) someText aNumber`? I don't know if it would be too magical to be easily understood.

But even if you think that is too magical, I'd prefer a simple and general mechanism for processing strings for different use-cases, that is usable by different libraries (cf. `-XOverloadedLabels`). But then you're back to running code at compile time, and basically back to TH. Hm...

Reply:
If we have built-in interpolation, it doesn't have to preclude the use of any other string-manipulation (formatting and interpolation) library. Satisfying the long tail of use cases is a great use of independent libraries, since there are so many mutually exclusive solutions to various cases. Maybe I'm just not smart enough to think of the most general solution that is still ergonomic and uncontroversial, but there's no reason to think it exists, and plenty to think it doesn't.

I think it would be fine to explicitly say that the goal for built-in interpolation is to be the best way to handle a useful and abundant subset of use cases. If it can be shown to be compatible with select other techniques used in the same module (which is intentionally a strong definition of "compatible"), perhaps that would be an even stronger and less controversial position.

Reply:
> > I'd like to know what the limiting principle is, because we could bundle everything with GHC in the name of convenience.
> 
> Well in this specific instance, we wouldn't "just" be bundling `formatting` into GHC. The benefit of adding this over just recommending `formatting` is because `formatting` wouldn't be able to take advantage of multiline strings (if that proposal is accepted). Compare:
> 
> ```haskell
> let s =
>       format
>       ("""
>       This is an example.
>       Here is a list of people and their ages:
>       Person named """ % text % " has age " % int % """.
>       Another person named """ % text % " has age " % int % """.
>       The end.
>       """)
>       name1 age1 name2 age2
> ```
> 
> vs
> 
> ```haskell
> let s =
>       s"""
>       This is an example.
>       Here is a list of people and their ages:
>       Person named ${name1} has age ${age1}.
>       Another person named ${name2} has age ${age2}.
>       The end.
>       """
> ```
> 
> Excusing the contrived example for a second, the second has much less glue getting in the way. If this isn't compelling to you, maybe we can just agree to disagree 🤷
> 
> > `formatting` is a lot better than interpolation
> 
> I guess I just personally disagree

These examples aren't equivalent, the former builds up a format and then applies it while the latter builds up a formatted string. The equivalent version of the latter would be ...named """ % name1 % """ has age.... Similarly I suppose you could have something like ${text} in your interpolated string that doesn't actually get treated as a fixed string, or have the overall result by polymorphic. 

For existing Haskell, suppose you instead used QualifiedDo and OverloadedStrings you could get this down to
```Haskell
F.do "This is an example."
     "Here is a list of people and their ages: "
     "Person named "%name1%" has age "%age1%"."
```
The cost becomes "% %" in place of ${ }. Conceivably you might even still have %text% for a delayed argument.
(You could use two different qualified do modules for auto ending and non auto ending lines.)